File: LLM/flash_attention.py
"""
FlashAttention-V1 分块流模拟 — 纯手工模拟 GPU SRAM 分块与 Online Softmax。

核心思想
--------
标准 Attention 物化完整 N×N 得分矩阵，显存 O(N²)。
FlashAttention 将 Q/K/V 切成小块在 SRAM 中计算，用 Online Softmax 跨块融合，
避免写出完整 attention matrix，显存 O(N)。

运行
----
    cd LLM
    python3 flash_attention.py
"""

from __future__ import annotations

import math

import torch
import torch.nn as nn
import torch.nn.functional as F


class FlashAttentionSimulator(nn.Module):
    """纯手工模拟 GPU SRAM 分块流的 FlashAttention-V1。"""

    def __init__(self, d_model: int, n_heads: int):
        super().__init__()
        self.n_heads = n_heads
        self.head_dim = d_model // n_heads

    def forward(self, Q: torch.Tensor, K: torch.Tensor, V: torch.Tensor,
                Br: int = 2, Bc: int = 2) -> torch.Tensor:
        """
        输入维度均为标准大模型格式：[Batch, num_heads, SeqLen, head_dim]

        Br: Q 行块大小（Query block rows）
        Bc: K/V 列块大小（Key/Value block columns）

        为了方便白板肉眼 Debug，默认将块大小设得极小 (Br=2, Bc=2)。
        """
        B, H, q_len, d = Q.shape
        _, _, kv_len, _ = K.shape

        # 1. 【静态开辟 HBM 空间】最终写回全局显存（HBM）的输出矩阵 O
        O = torch.zeros_like(Q)

        # 2. 【外层：遍历 Q 的行块】每个 Q_i 独立维护 Online Softmax 状态
        for i in range(0, q_len, Br):
            # 模拟搬运当前 Q_i 到 SRAM
            Q_i = Q[:, :, i : i + Br, :]
            O_i = torch.zeros_like(Q_i)

            # 模拟在线局部寄存器状态 (Per-row-block Stats)
            # 在真实 CUDA 中，m_old / d_old 常驻 GPU 寄存器，跨 K/V 列块滚动更新
            m_old = torch.full((B, H, Br, 1), float("-inf"), device=Q.device, dtype=Q.dtype)
            d_old = torch.zeros((B, H, Br, 1), device=Q.device, dtype=Q.dtype)

            # 3. 【内层：遍历 K, V 的列块】模拟 HBM -> SRAM 搬运
            for j in range(0, kv_len, Bc):
                K_j = K[:, :, j : j + Bc, :]  # [B, H, Bc, d]
                V_j = V[:, :, j : j + Bc, :]

                # SRAM 内部寄存器级乘法: S_ij [B, H, Br, Bc]
                S_ij = torch.matmul(Q_i, K_j.transpose(-2, -1)) / math.sqrt(d)

                # --- Online Softmax 核心 ---
                m_block, _ = torch.max(S_ij, dim=-1, keepdim=True)
                exp_S = torch.exp(S_ij - m_block)
                d_block = torch.sum(exp_S, dim=-1, keepdim=True)

                # -------------------------------------------------------------
                # 【工业级高度融合：Online Softmax 跨块对齐公式】
                # -------------------------------------------------------------
                # 1. 动态融合、更新最新的全局最大值
                m_new = torch.max(m_old, m_block)

                # 2. 动态缩放并累加分母
                d_new = d_old * torch.exp(m_old - m_new) + d_block * torch.exp(m_block - m_new)

                # 3. 校正旧 Output 并融入新 V_j 块的贡献
                O_i_corrected = O_i * (d_old * torch.exp(m_old - m_new) / (d_new + 1e-9))
                O_i = O_i_corrected + torch.matmul(
                    exp_S * torch.exp(m_block - m_new), V_j
                ) / (d_new + 1e-9)

                m_old = m_new
                d_old = d_new

            # 4. 将最终 O_i 刷写回 HBM
            O[:, :, i : i + Br, :] = O_i

        return O


def naive_attention(Q: torch.Tensor, K: torch.Tensor, V: torch.Tensor) -> torch.Tensor:
    """标准 Scaled Dot-Product Attention，用于数值对照。"""
    d = Q.shape[-1]
    scores = torch.matmul(Q, K.transpose(-2, -1)) / math.sqrt(d)
    attn = F.softmax(scores, dim=-1)
    return torch.matmul(attn, V)


def demo() -> None:
    torch.manual_seed(0)
    device = "cuda" if torch.cuda.is_available() else "cpu"

    B, H, seq_len, d = 1, 2, 6, 4
    Br, Bc = 2, 2

    Q = torch.randn(B, H, seq_len, d, device=device)
    K = torch.randn(B, H, seq_len, d, device=device)
    V = torch.randn(B, H, seq_len, d, device=device)

    model = FlashAttentionSimulator(d_model=H * d, n_heads=H).to(device)

    print(f"--- FlashAttention Simulator (Br={Br}, Bc={Bc}, seq={seq_len}) ---")
    with torch.no_grad():
        O_flash = model(Q, K, V, Br=Br, Bc=Bc)
        O_naive = naive_attention(Q, K, V)

    max_diff = (O_flash - O_naive).abs().max().item()
    print(f"Output shape: {tuple(O_flash.shape)}")
    print(f"Max |flash - naive|: {max_diff:.2e}")
    assert max_diff < 1e-4, f"numerical mismatch: {max_diff}"
    print("FlashAttention matches naive attention.")


if __name__ == "__main__":
    demo()

File: LLM/flash_attention.cpp
// FlashAttention-V1 block-tile simulator — plain C++ (no PyTorch).
//
// Mirrors LLM/flash_attention.py: tile Q/K/V into SRAM-sized blocks and fuse
// Online Softmax across K/V column blocks without materializing N×N scores.
//
// Whiteboard talking points:
// - Naive attention writes full score matrix -> O(N²) HBM traffic.
// - FlashAttention keeps m/d accumulators in registers per Q row block.
// - Outer loop over Q rows, inner loop over K/V columns (standard V1 tiling).
// - Production kernels fuse even further and use warp-level primitives.

#include <algorithm>
#include <cassert>
#include <cmath>
#include <cstddef>
#include <iostream>
#include <limits>
#include <random>
#include <stdexcept>
#include <vector>

struct Tensor4D {
    int batch = 0;
    int heads = 0;
    int seq = 0;
    int head_dim = 0;
    std::vector<float> data;

    float& at(int b, int h, int s, int d) {
        return data[((b * heads + h) * seq + s) * head_dim + d];
    }

    float at(int b, int h, int s, int d) const {
        return data[((b * heads + h) * seq + s) * head_dim + d];
    }
};

static Tensor4D zeros_like(const Tensor4D& ref) {
    return Tensor4D{ref.batch, ref.heads, ref.seq, ref.head_dim,
                    std::vector<float>(ref.data.size(), 0.0f)};
}

static Tensor4D flash_attention(const Tensor4D& Q, const Tensor4D& K, const Tensor4D& V,
                                int Br, int Bc) {
    if (Q.batch != K.batch || Q.heads != K.heads || Q.head_dim != K.head_dim ||
        K.batch != V.batch || K.heads != V.heads || K.head_dim != V.head_dim) {
        throw std::invalid_argument("Q/K/V shape mismatch");
    }

    const int B = Q.batch;
    const int H = Q.heads;
    const int q_len = Q.seq;
    const int kv_len = K.seq;
    const int d = Q.head_dim;
    const float scale = 1.0f / std::sqrt(static_cast<float>(d));
    const float neg_inf = -std::numeric_limits<float>::infinity();

    Tensor4D O = zeros_like(Q);

    for (int i = 0; i < q_len; i += Br) {
        const int br = std::min(Br, q_len - i);

        std::vector<float> O_i(static_cast<size_t>(B) * H * br * d, 0.0f);
        std::vector<float> m_old(static_cast<size_t>(B) * H * br, neg_inf);
        std::vector<float> d_old(static_cast<size_t>(B) * H * br, 0.0f);

        auto o_idx = [&](int b, int h, int r, int dd) {
            return ((b * H + h) * br + r) * d + dd;
        };
        auto stat_idx = [&](int b, int h, int r) { return (b * H + h) * br + r; };

        for (int j = 0; j < kv_len; j += Bc) {
            const int bc = std::min(Bc, kv_len - j);

            for (int b = 0; b < B; ++b) {
                for (int h = 0; h < H; ++h) {
                    for (int r = 0; r < br; ++r) {
                        const int si = stat_idx(b, h, r);

                        std::vector<float> scores(static_cast<size_t>(bc), 0.0f);
                        for (int c = 0; c < bc; ++c) {
                            float dot = 0.0f;
                            for (int dd = 0; dd < d; ++dd) {
                                dot += Q.at(b, h, i + r, dd) * K.at(b, h, j + c, dd);
                            }
                            scores[static_cast<size_t>(c)] = dot * scale;
                        }

                        const float m_block =
                            *std::max_element(scores.begin(), scores.end());
                        std::vector<float> exp_s(static_cast<size_t>(bc), 0.0f);
                        float d_block = 0.0f;
                        for (int c = 0; c < bc; ++c) {
                            exp_s[static_cast<size_t>(c)] =
                                std::exp(scores[static_cast<size_t>(c)] - m_block);
                            d_block += exp_s[static_cast<size_t>(c)];
                        }

                        const float m_new = std::max(m_old[static_cast<size_t>(si)], m_block);
                        const float alpha = std::exp(m_old[static_cast<size_t>(si)] - m_new);
                        const float beta = std::exp(m_block - m_new);
                        const float d_new =
                            d_old[static_cast<size_t>(si)] * alpha + d_block * beta;
                        const float corr =
                            (d_old[static_cast<size_t>(si)] * alpha) / (d_new + 1e-9f);

                        for (int dd = 0; dd < d; ++dd) {
                            float sum = O_i[o_idx(b, h, r, dd)] * corr;
                            for (int c = 0; c < bc; ++c) {
                                sum += (exp_s[static_cast<size_t>(c)] * beta) *
                                       V.at(b, h, j + c, dd) / (d_new + 1e-9f);
                            }
                            O_i[o_idx(b, h, r, dd)] = sum;
                        }

                        m_old[static_cast<size_t>(si)] = m_new;
                        d_old[static_cast<size_t>(si)] = d_new;
                    }
                }
            }
        }

        for (int b = 0; b < B; ++b) {
            for (int h = 0; h < H; ++h) {
                for (int r = 0; r < br; ++r) {
                    for (int dd = 0; dd < d; ++dd) {
                        O.at(b, h, i + r, dd) = O_i[o_idx(b, h, r, dd)];
                    }
                }
            }
        }
    }

    return O;
}

static std::vector<float> softmax_row(const std::vector<float>& row) {
    const float max_v = *std::max_element(row.begin(), row.end());
    std::vector<float> out(row.size());
    float sum = 0.0f;
    for (size_t i = 0; i < row.size(); ++i) {
        out[i] = std::exp(row[i] - max_v);
        sum += out[i];
    }
    for (float& v : out) {
        v /= sum;
    }
    return out;
}

static Tensor4D naive_attention(const Tensor4D& Q, const Tensor4D& K, const Tensor4D& V) {
    const int B = Q.batch;
    const int H = Q.heads;
    const int q_len = Q.seq;
    const int kv_len = K.seq;
    const int d = Q.head_dim;
    const float scale = 1.0f / std::sqrt(static_cast<float>(d));

    Tensor4D O = zeros_like(Q);

    for (int b = 0; b < B; ++b) {
        for (int h = 0; h < H; ++h) {
            for (int i = 0; i < q_len; ++i) {
                std::vector<float> scores(static_cast<size_t>(kv_len), 0.0f);
                for (int j = 0; j < kv_len; ++j) {
                    float dot = 0.0f;
                    for (int dd = 0; dd < d; ++dd) {
                        dot += Q.at(b, h, i, dd) * K.at(b, h, j, dd);
                    }
                    scores[static_cast<size_t>(j)] = dot * scale;
                }
                const std::vector<float> attn = softmax_row(scores);
                for (int dd = 0; dd < d; ++dd) {
                    float sum = 0.0f;
                    for (int j = 0; j < kv_len; ++j) {
                        sum += attn[static_cast<size_t>(j)] * V.at(b, h, j, dd);
                    }
                    O.at(b, h, i, dd) = sum;
                }
            }
        }
    }
    return O;
}

static float max_abs_diff(const Tensor4D& a, const Tensor4D& b) {
    float max_diff = 0.0f;
    for (size_t i = 0; i < a.data.size(); ++i) {
        max_diff = std::max(max_diff, std::abs(a.data[i] - b.data[i]));
    }
    return max_diff;
}

static Tensor4D random_tensor(int batch, int heads, int seq, int head_dim, std::mt19937& rng) {
    std::normal_distribution<float> dist(0.0f, 1.0f);
    Tensor4D t{batch, heads, seq, head_dim,
               std::vector<float>(static_cast<size_t>(batch) * heads * seq * head_dim)};
    for (float& v : t.data) {
        v = dist(rng);
    }
    return t;
}

int main() {
    constexpr int Br = 2;
    constexpr int Bc = 2;

    std::mt19937 rng(0);
    const Tensor4D Q = random_tensor(1, 2, 6, 4, rng);
    const Tensor4D K = random_tensor(1, 2, 6, 4, rng);
    const Tensor4D V = random_tensor(1, 2, 6, 4, rng);

    const Tensor4D O_flash = flash_attention(Q, K, V, Br, Bc);
    const Tensor4D O_naive = naive_attention(Q, K, V);

    const float max_diff = max_abs_diff(O_flash, O_naive);
    std::cout << "--- FlashAttention Simulator (Br=" << Br << ", Bc=" << Bc << ", seq=6) ---\n";
    std::cout << "Max |flash - naive|: " << max_diff << "\n";

    assert(max_diff < 1e-3f);
    std::cout << "flash_attention: all tests passed\n";
    return 0;
}
